atlasdb-impl-shared/src/main/java/com/palantir/lock/watch/LockWatchReferencesVisitor.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.lock.watch.LockWatchReferences.EntireTable;
+import com.palantir.lock.watch.LockWatchReferences.ExactCell;
+import com.palantir.lock.watch.LockWatchReferences.ExactRow;
+import com.palantir.lock.watch.LockWatchReferences.RowPrefix;
+import com.palantir.lock.watch.LockWatchReferences.RowRange;
+import java.util.Optional;
+
+public final class LockWatchReferencesVisitor implements LockWatchReferences.Visitor<Optional<TableReference>> {
+    public static final LockWatchReferencesVisitor INSTANCE = new LockWatchReferencesVisitor();
+
+    @Override
+    public Optional<TableReference> visit(EntireTable reference) {
+        return Optional.of(TableReference.createFromFullyQualifiedName(reference.qualifiedTableRef()));
+    }
+
+    @Override
+    public Optional<TableReference> visit(RowPrefix reference) {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TableReference> visit(RowRange reference) {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TableReference> visit(ExactRow reference) {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TableReference> visit(ExactCell reference) {
+        return Optional.empty();
+    }
```

Comment:
What's the plan for the other types? They _do_ have table references too. I don't know exactly how this is going to be used so leaving this up to you

Reply:
Technically speaking, they do have table references, but I don't want to think that I'm watching a whole table when actually it is just an `ExactCell`. I think, for now, that these should just throw, and one day we'll implement more specific watching.